### PR TITLE
First draft for user code introspection into components

### LINKED ISSUE #1429: How to use Autosklearn with Multi-Class Multi-Label Classification?
I have been trying to use AutoSklearn with Multi-class classification

so my labels are like this

0   1   2    3    4 ...  200
1   0   1   1     1 ...   1
0   1   0   0     1 ...   0
1   0   0   1     0 ...   0
1   1   0   1     0 ...   1
0   1   1   0     1 ...   0
1   1   1   0     0 ...   1
1   0   1   0     1 ...   0

I used this code

`
y = y[:, (65,67,54,133,122,63,102,105,39)]
X = df.drop(Code, axis=1, errors='ignore')
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

automl = autosklearn.classification.AutoSklearnClassifier(
include={'feature_preprocessor': ["no_preprocessing"], 
 },
exclude={ 'classifier': ['random_forest']},
time_left_for_this_task=60*5,
per_run_time_limit=60*1,
memory_limit = 1024 * 10,
n_jobs=-1,
metric=autosklearn.metrics.f1_macro,
        )

`

but now  I want to train Autosklearn on Multi-class Multi-label classification

Which method of these shall i use?

**1-** 

```
clf = OneVsRestClassifier(automl, n_jobs=-1)
clf.fit(X_train, y_train)

```

**2-** 
```

clf = automl
clf.fit(X_train, y_train)
```

**3-**

I have to loop one class at a time and use 

```
clf = automl
clf.fit(X_train, y_train)
```

so it will be like 

```
for i in (65,67,54,133,122,63,102,105,39):
       y = z[:, i]
       X = df.drop(Code, axis=1, errors='ignore')
       X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
      automl = autosklearn.classification.AutoSklearnClassifier(
      include={'feature_preprocessor': ["no_preprocessing"], 
       },
      exclude={ 'classifier': ['random_forest']},
      time_left_for_this_task=60*5,
      per_run_time_limit=60*1,
      memory_limit = 1024 * 10,
      n_jobs=1,
      metric=autosklearn.metrics.f1_macro,
              )


      clf = automl
      clf.fit(X_train, y_train)
```

so I get a different model for each label?

### PR DESCRIPTION
**Pull request by eddiebergman**
This is a first draft in an attempt to address #1429. Sample output:

```python
from autosklearn.info import classifiers

name, info = next(iter(classifiers().items()))

print(name)
"adaboost"

print(info)
ClassifierInfo(type=<class 'autosklearn.pipeline.components.classification.adaboost.AdaboostClassifier'>, name='AdaBoost Classifier', shortname='AB', output_kind='predictions', supported_inputs=['dense', 'sparse', 'unsigned data'], deterministic=True, handles_binary=True, handles_multiclass=True, handles_multilabel=False)
```

The `classifiers()` returns `dict[str, ClassifierInfo]`

```python
classifiers()

{'adaboost': ClassifierInfo(type=<class 'autosklearn.pipeline.components.classification.adaboost.AdaboostClassifier'>, name='AdaBoost Classifier', shortname='AB', output_kind='predictions', supported_inputs=['dense', 'sparse', 'unsigned data'], deterministic=True, handles_binary=True, handles_multiclass=True, handles_multilabel=False),
 'bernoulli_nb': ClassifierInfo(type=<class 'autosklearn.pipeline.components.classification.bernoulli_nb.BernoulliNB'>, name='Bernoulli Naive Bayes classifier', shortname='BernoulliNB', output_kind='predictions', supported_inputs=['dense', 'sparse', 'unsigned data'], deterministic=True, handles_binary=True, handles_multiclass=True, handles_multilabel=True),
 'decision_tree': ClassifierInfo(type=<class 'autosklearn.pipeline.components.classification.decision_tree.DecisionTree'>, name='Decision Tree Classifier', shortname='DT', output_kind='predictions', supported_inputs=['dense', 'sparse', 'unsigned data'], deterministic=True, handles_binary=True, handles_multiclass=True, handles_multilabel=True),
 'extra_trees': ClassifierInfo(type=<class 'autosklearn.pipeline.components.classification.extra_trees.ExtraTreesClassifier'>, name='Extra Trees Classifier', shortname='ET', output_kind='predictions', supported_inputs=['dense', 'sparse', 'unsigned data'], deterministic=True, handles_binary=True, handles_multiclass=True, handles_multilabel=True),
 'gaussian_nb': ClassifierInfo(type=<class 'autosklearn.pipeline.components.classification.gaussian_nb.GaussianNB'>, name='Gaussian Naive Bayes classifier', shortname='GaussianNB', output_kind='predictions', supported_inputs=['dense', 'unsigned data'], deterministic=True, handles_binary=True, handles_multiclass=True, handles_multilabel=True),
 'gradient_boosting': ClassifierInfo(type=<class 'autosklearn.pipeline.components.classification.gradient_boosting.GradientBoostingClassifier'>, name='Gradient Boosting Classifier', shortname='GB', output_kind='predictions', supported_inputs=['dense', 'unsigned data'], deterministic=True, handles_binary=True, handles_multiclass=True, handles_multilabel=False),
 'k_nearest_neighbors': ClassifierInfo(type=<class 'autosklearn.pipeline.components.classification.k_nearest_neighbors.KNearestNeighborsClassifier'>, name='K-Nearest Neighbor Classification', shortname='KNN', output_kind='predictions', supported_inputs=['dense', 'sparse', 'unsigned data'], deterministic=True, handles_binary=True, handles_multiclass=True, handles_multilabel=True),
 'lda': ClassifierInfo(type=<class 'autosklearn.pipeline.components.classification.lda.LDA'>, name='Linear Discriminant Analysis', shortname='LDA', output_kind='predictions', supported_inputs=['dense', 'unsigned data'], deterministic=True, handles_binary=True, handles_multiclass=True, handles_multilabel=True),
 'liblinear_svc': ClassifierInfo(type=<class 'autosklearn.pipeline.components.classification.liblinear_svc.LibLinear_SVC'>, name='Liblinear Support Vector Classification', shortname='Liblinear-SVC', output_kind='predictions', supported_inputs=['sparse', 'dense', 'unsigned data'], deterministic=False, handles_binary=True, handles_multiclass=True, handles_multilabel=True),
 'libsvm_svc': ClassifierInfo(type=<class 'autosklearn.pipeline.components.classification.libsvm_svc.LibSVM_SVC'>, name='LibSVM Support Vector Classification', shortname='LibSVM-SVC', output_kind='predictions', supported_inputs=['dense', 'sparse', 'unsigned data'], deterministic=True, handles_binary=True, handles_multiclass=True, handles_multilabel=False),
 'mlp': ClassifierInfo(type=<class 'autosklearn.pipeline.components.classification.mlp.MLPClassifier'>, name='Multilayer Percepton', shortname='MLP', output_kind='predictions', supported_inputs=['dense', 'sparse', 'unsigned data'], deterministic=True, handles_binary=True, handles_multiclass=True, handles_multilabel=True),
 'multinomial_nb': ClassifierInfo(type=<class 'autosklearn.pipeline.components.classification.multinomial_nb.MultinomialNB'>, name='Multinomial Naive Bayes classifier', shortname='MultinomialNB', output_kind='predictions', supported_inputs=['dense', 'sparse', 'signed data'], deterministic=True, handles_binary=True, handles_multiclass=True, handles_multilabel=True),
 'passive_aggressive': ClassifierInfo(type=<class 'autosklearn.pipeline.components.classification.passive_aggressive.PassiveAggressive'>, name='Passive Aggressive Classifier', shortname='PassiveAggressive Classifier', output_kind='predictions', supported_inputs=['dense', 'sparse', 'unsigned data'], deterministic=True, handles_binary=True, handles_multiclass=True, handles_multilabel=True),
 'qda': ClassifierInfo(type=<class 'autosklearn.pipeline.components.classification.qda.QDA'>, name='Quadratic Discriminant Analysis', shortname='QDA', output_kind='predictions', supported_inputs=['dense', 'unsigned data'], deterministic=True, handles_binary=True, handles_multiclass=True, handles_multilabel=True),
 'random_forest': ClassifierInfo(type=<class 'autosklearn.pipeline.components.classification.random_forest.RandomForest'>, name='Random Forest Classifier', shortname='RF', output_kind='predictions', supported_inputs=['dense', 'sparse', 'unsigned data'], deterministic=True, handles_binary=True, handles_multiclass=True, handles_multilabel=True),
 'sgd': ClassifierInfo(type=<class 'autosklearn.pipeline.components.classification.sgd.SGD'>, name='Stochastic Gradient Descent Classifier', shortname='SGD Classifier', output_kind='predictions', supported_inputs=['dense', 'sparse', 'unsigned data'], deterministic=True, handles_binary=True, handles_multiclass=True, handles_multilabel=False)}
```

The same exists for all `classifiers`, `regressors`, `data_preprocessors` and `feature_preprocessors`

```python
from autosklearn.info import components

# Same as the classifiers above
all_components = components()
classifiers = all_components.classfiers
feature_preprocessors = all_components.feature_preprocessors
```
---
### Issues
* Data preprocessors are their own beast, there's technically only one datapreprocessor component `FeatTypeSplit`
    * We've never come up with a nice solution to this as the pipeline is not flexible enough to really represent what we want

* There's a lot of duplication existing already, i.e. `handles_spares = True` and `input = (SPARSE, ...)` in `get_properties()`.
    * The dataclasses defined could just be used directly in each component. It makes it type safe, use `.` notation instead of `["this"]`.
    * Has the added benefit of really simplifying the stuff in `autosklearn.info` too.
    * Also means one source of truth for what's allowed in those properties

* I would prefer to do what's below long term, it's a bit nicer to look at. The problem now is that if a user adds a custom component, I would like it to show up when they do `components`. This is fixable, just noting that it could be changed.
```python
from autosklearn.info import components


# Current requires calling
for name, info in components().classfiers.items():
    # do stuff

# Note how components doesn't need to be called
for name, info in components.classfiers.items():
    # do stuff
```
---
* Closes #1429

### CODE REVIEW

File: autosklearn/pipeline/components/regression/sgd.py [hunk 1]
```diff
@@ -179,7 +179,6 @@ def get_properties(dataset_properties=None):
             "handles_multilabel": False,
             "handles_multioutput": False,
             "is_deterministic": True,
-            "handles_sparse": True,
```
Comment (KEggensperger) [inline autosklearn/pipeline/components/regression/sgd.py:182]: Why do we remove this here?
